File: medium/208.implement-trie-prefix-tree.py
"""
* .insert(cat)
c, .insert(at)
a, .insert(t)
t, .insert()
"""


class Trie:

    def __init__(self, val:str = None):
        self.val = val
        #print(val)
        self.children = {}
        self.is_terminal = False
        

    def insert(self, word: str) -> None:
        if len(word) == 0:
            self.is_terminal = True
            return
        ch = word[0]
        node = None
        if ch in self.children:
            node = self.children[ch]
        else:
            node = Trie(ch)
            self.children[ch] = node
        node.insert(word[1:])
        #print(word[1:])
        
        

    def search(self, word: str) -> bool:
        if len(word) == 0:
            return self.is_terminal
        ch = word[0]
        #print(ch, " in ", self.children)
        if ch in self.children:
            return self.children[ch].search(word[1:])
        return False
        

    def startsWith(self, prefix: str) -> bool:
        if len(prefix) == 0:
            return True
        ch = prefix[0]
        if ch in self.children:
            return self.children[ch].startsWith(prefix[1:])
        return False
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: medium/208.implement-trie-prefix-tree.rs
use std::collections::HashMap;
/////////Leetcode
#[derive(Default)]
struct Trie {
    child: HashMap<char, Box<Trie>>,
    is_terminal: bool,

}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Trie {

    fn new() -> Self {
        Default::default()
        
    }
    
    fn insert(&mut self, word: String) {
        let mut p = self;
        for c in word.chars() {
            p = p.child.entry(c).or_insert(Default::default());
        }
        p.is_terminal = true;
    }
    
    fn search(&self, word: String) -> bool {
        let mut p = self;
        for c in word.chars() {
            if let Some(next) = p.child.get(&c) {
                p = next;
            } else {
                return false;
            }
        }
        p.is_terminal
    }
    
    fn starts_with(&self, prefix: String) -> bool {
        let mut p = self;
        for c in prefix.chars() {
            if let Some(next) = p.child.get(&c) {
                p = next;
            } else {
                return false;
            }
        }
        true
    }
}